onadata/libs/utils/logger_tools.py

```diff
@@ -241,6 +241,9 @@ def save_attachments(xform, instance, media_files):
                 mimetype=content_type,
                 name=filename,
                 extension=extension)
+    Attachment.soft_delete(Attachment.objects.filter(instance=instance).filter(
+        ~Q(name__in=instance.get_expected_media())))
+
```

Comment:
Why is it relevant to delete attachments every time we receive a submission?

Reply:
This allows us to apply soft deletion action against the replaced attachments, if any.
At the moment, if a user edits a submission with 4 attachments and replaces 1 of the attachments with a new media file, the replaced attachment is retained as a valid attachment for that instance and not really deleted, rather a new attachment is created for that submission

Having this filter will allow us know and delete attachments no longer prevalent for that instance.

Reply:
I would recommend we add an optional field such that we only delete on edits. We can apply true to this field at https://github.com/onaio/onadata/blob/df3d077d6467e07b1d69003de0efc04a442aee74/onadata/libs/utils/logger_tools.py#L320.

```python
def save_attachments(xform, instance, media_files, remove_deleted_media=False):
...
        
    if remove_deleted_media:
       Attachment.soft_delete(instance.attachments.filter(~Q(name__in=instance.get_expected_media())))
```

Reply:
Sure. I have included this optional field.